.github/workflows/ci.yml

```diff
@@ -41,7 +47,15 @@ jobs:
         if: success() && matrix.os == 'ubuntu-latest'
         run: npm install --prefix web-playground
 
-      - name: Build web playground frontend
+      - name: Start web playground frontend
+        if: success() && matrix.os == 'ubuntu-latest'
+        run: npm start --prefix web-playground &
+
+      - name: Test web playground frontend
+        if: success() && matrix.os == 'ubuntu-latest'
+        run: npm run e2e --prefix web-playground
```

Comment:
Does the latter task depend on the first reaching a state where it's ready? Can it fail if the first is still pending?

Reply:
No, I think it's up and running when the command ends.

Reply:
Cool :+1: Fair enough. My worry was that sending it to the background with `&` wouldn't guarantee that.

Reply:
Oh, right, I see now. Yes, you are right in principle. In practice, it seems to work. Do you want to wait and see if it's flaky, or do you know a good way to improve it now?

Reply:
Another way, which is either genius or super dirty, is to defer the testing until after the following build step.

Reply:
I think that's both super dirty and genius simultaneously :smile: A `curl` loop is probably safest and expresses the intention clearly... Let me add that :+1:

Reply:
Good!

Now that I think about it, it might be best to move the frontend build step ahead of all of this. So if there actually is something wrong with the app, we get an error about that and not just about the app refusing to start.

Reply:
I added the `curl` check in f884d2b

Reply:
Yep :disappointed: I don't know why. I've turned off the redirect to `/dev/null` to see if that reveals anything.

Reply:
My machines say this:

```
curl: option --connect-timeout=10: is unknown
```

Reply:
Yep, that's the problem. Apparently `curl` doesn't like long options with equal signs. #TIL

Reply:
Interesting. This one doesn't wait at all:

```bash
curl --connect-timeout 10 http://localhost:3000
curl: (7) Failed to connect to localhost port 3000 after 0 ms: Connection refused
```

Reply:
I'm just trawling through the manual. I think it should be `--max-time` instead...

Reply:
That one also fails right away. Apparently these two are only max limits, not min.

Reply:
...Sigh. Nope, that doesn't work either.

Reply:
I was hoping `curl` had a nicer way, but it looks like the loop is just going to be easier :+1:

Reply:
If anything, I should be apologising for dragging you into this 🙂

But this is definitely an improvement. Now we get to see the full output of the start command, so it will be easy to see if it has failed, and why.